plugins/obs-filters, UI: Add support for N sidechain ducking sources for the compressor filter

Avuxo opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
This patch adds in the ability to use N ducking sources in the compressor filter. 

### Motivation and Context
In broadcasting settings, there are often times when you have more than one ducking source into a compressor. A common example is multiple commentators commenting over a background audio track. Right now this requires expensive extra steps involved in the pipeline. Further, this is a much more obvious solution requiring less tinkering.

### How Has This Been Tested?
This has been tested in the following way:
1. create a compression target media source with the following params:
   i. threshold at min and release at max to make the effect obvious
2. Create 1 input source, then 2, then 3, then 4 and for each:
   i. Add it as a compression sidechain source to compression target
   ii. mute and unmute each individually, then all together
   iii. Pause and unpause each individually then all together.

The effect should be very obvious in the levels view if threshold is set at minimum and release at maximum.
For best effect, use both ducking sources and target with constant audio peaking the level of its given channel.

Tested on macOS (Apple silicon 14.5), Windows 11 (x86_64), and Ubuntu 24.04 (x86_64)

### Types of changes
New feature (non-breaking change which adds functionality)
Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Avuxo committed:
UI: Add editable list widget for sources

Avuxo committed:
plugins/obs-filters: Refactor compressor sidechain out into struct

Avuxo committed:
plugins/obs-filters: Refactor compressor to support >1 sidechains

Avuxo committed:
plugins/obs-filters: Refactored compressor properly destroyed

Avuxo committed:
plugins/obs-filters: Fix pause logic in compressor refactor

Avuxo committed:
plugins/obs-filters: Optimize sidechain post-refactor

Warchamp7 commented:
The PR and commit title currently don't adhere to our contribution guidelines. Can you also add screenshots or a video showing the changed behaviour here?

As it stands I'm currently leaning towards closing this as advanced audio control of this manner is something we generally feel is better left to dedicated audio software or hardware.

Avuxo commented:
@Warchamp7 
Thank you for the comment.

Regarding the behavioral change, I have added an editable list of sources to the compressor filter such that you can have as many sidechain/ducking sources as you want. In short, this means that you can have any number of sources that can independently trigger the compressor.

Consider a sports broadcast where you might have two or three sports commentators speaking over game audio. In this case, you want to apply compression to the game audio when any of the sports commentators activates their mic over a certain threshold. This is our use case for OBS.
![image](https://github.com/user-attachments/assets/0f5e542a-83c2-401b-b93c-6b2dd9f9dc67)

I would argue that this is, regardless of lines changed, a relatively minor behavioral change to the compressor filter and fits within the purview of OBS itself. Rather than being a drop-down, it is now an editable list from which you can select up to N sidechain/ducking sources (triggers) for compression.

I apologize for mistakes in the title or commit history. I will promptly address any concerns

Avuxo commented:
I have updated the commit headings to hopefully be within guidelines. I apologize.

Avuxo committed:
plugins/obs-filters: Include compressor analysis in critical section

After using this in production, it became clear that without the lock
containing the entire sidechain analysis, it would be possible for the
user to, at an inopportune time, modify the compressor settings (even
those unrelated to sidechains), causing the references to be reset
mid-analysis, and thus a stale reference would be dereferenced and would
cause a crash.

This bug can be reproduced pre-commit by fiddling a lot with the the
compressor filter while it is active. This fix has been tested in
production for some of our live events now.

Avuxo commented:
I have included a fix to a bug that we encountered in production that would result in an OBS crash when using this patch. Details in commit message. It is clear that without the lock containing the entire sidechain analysis, it would be possible for the user to, at an inopportune time, modify the compressor settings (even those unrelated to sidechains), causing the references to be reset mid-analysis, and thus a stale reference would be dereferenced and would cause a crash.

The current version of the patch has been tested in production successfully and I ran it through a more thorough test of invalidating and revalidating references. A future possible improvement would be to more intelligently detect when a sidechain _needs_ to have its weak reference updated.

Avuxo committed:
Merge branch 'master' into multichain-2

Avuxo committed:
Merge branch 'master' into multichain-2
